source/CIS/cis-5.rst

```diff
@@ -0,0 +1,621 @@
+  Only interact with a CIS-2 token contract or value its recorded token balance if you checked its smart
+  contract logic or reasonable social reputation is given to the project/CIS-2 token contract.
+
+Parameter
+~~~~~~~~~
+
+The parameter is the :ref:`CIS-2-functions-transfer-receive-hook-parameter` (``OnReceivingCis2Params``) and the
+``data`` field of the ``OnReceivingCis2Params`` SHALL encode a ``PublicKeyEd25519``.
+
+See the serialization rules in :ref:`CIS-2-functions-transfer-receive-hook-parameter`
+and the serialization rules in :ref:`CIS-5-PublicKeyEd25519`.
+
+Requirements
+~~~~~~~~~~~~
+
+- The function SHOULD check that a contract is the caller since only a contract can implement a receive hook mechanism.
+
+.. _CIS-5-functions-withdrawCcd:
+
+``withdrawCcd``
+^^^^^^^^^^^^^^^
+
+The function executes a list of token withdrawals of CCDs to native accounts and/or smart contracts out of the smart contract wallet.
+When transferring CCD to a contract address, a CCD receive hook function MUST be triggered.
+
+Parameter
+~~~~~~~~~
+
+The parameter is a list of withdrawals.
+
+It is serialized as: 2 bytes representing the number of withdrawals (``n``) followed by the bytes for this number of ``withdrawals``.
+
+Each ``WithdrawBatchCcdAmount`` is serialized as: a :ref:`CIS-5-PublicKeyEd25519` (``signer``), a :ref:`CIS-5-SignatureEd25519` (``signature``), and  a ``WithdrawMessageCcdAmount`` (``message``).
+
+Each ``WithdrawMessageCcdAmount`` is serialized as: an :ref:`CIS-5-EntrypointName` (``entryPoint``), a :ref:`CIS-5-TimeStamp` (``expiryTime``), a :ref:`CIS-5-Nonce` (``nonce``), a :ref:`CIS-5-PublicKeyEd25519` (``serviceFeeRecipient``), a :ref:`CIS-5-CCDAmount` (``serviceFee``), 2 bytes representing the number of simple withdraws (``m``) followed by the bytes for this number of simple withdraws (``simpleWithdraws``).
+
+Each ``WithdrawCcdAmount`` is serialized as: the receiving address :ref:`CIS-2-Receiver` (``to``), the :ref:`CIS-5-CCDAmount` (``withdrawAmount``), and some additional data :ref:`CIS-2-AdditionalData` (``data``)::
+
+  WithdrawCcdAmount ::=  (to: Receiver) (withdrawAmount: CCDAmount) (data: AdditionalData)
+
+  WithdrawMessageCcdAmount ::= (entryPoint: EntrypointName) (expiryTime: TimeStamp) (nonce: u64) (serviceFeeRecipient: PublicKeyEd25519) (serviceFee: CCDAmount) (m: Byte²) (simpleWithdraws: WithdrawCcdAmountᵐ)
+
+  WithdrawBatchCcdAmount ::= (signer: PublicKeyEd25519) (signature: SignatureEd25519) (message: WithdrawMessageCcdAmount)
+
+  WithdrawParameterCcdAmount ::= (n: Byte²) (withdrawals: WithdrawBatchCcdAmountⁿ)
+
+.. _CIS-5-functions-transfer-ccd-receive-hook-parameter:
+
+CCD Receive hook parameter
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The parameter for the CCD receive hook function contains some additional data bytes.
+
+It is serialized as: some additional data :ref:`CIS-2-AdditionalData` (``data``)::
+
+  CCDReceiveHookParameter ::= (data: AdditionalData)
```

Comment:
I was thinking about it but it would prevent the composability with other DEFI smart contract protocols. E.g. If people want to trade/deposit their smart contract wallet funds into a DEX/staking protocol they can do it as part of the `withdraw` function now by specifying in the `data` whatever these DEFI protocol input parameters are. If we include the `public_key` here, people would need at least 2 transactions to interact with DEFI protocols.

As defined now, I see the `public_key` as optional. It can still be added to the additional data if the receiving smart contract expects/wants to know the public key, but we don't force it.

Reply:
It will only work if that protocol serializes the parameter similar to `AdditionalData` (2 bytes length prefix).
So if the idea is to be able to interact with any smart contract, we probably should not use `AdditionalData` as the type.